services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloaderSubscriber.java

```diff
@@ -27,86 +27,115 @@
         this.s3 = s3;
         this.getObjectRequest = getObjectRequest;
-        this.responseFuture = new CompletableFuture<>();
     }
 
     @Override
     public void onSubscribe(Subscription s) {
-        if (this.subscription == null) {
-            this.subscription = s;
+        if (this.subscription != null) {
+            s.cancel();
+            return;
         }
-        s.request(1);
+        this.subscription = s;
+        this.subscription.request(1);
     }
 
     @Override
     public void onNext(AsyncResponseTransformer<GetObjectResponse, GetObjectResponse> asyncResponseTransformer) {
-        int part = currentPart.incrementAndGet();
-        if (totalPartKnown.get()) {
-            log.info(() -> String.format("total part: %s, current part: %s", totalParts.get(), part));
-        } else {
-            log.info(() -> String.format("part %s", part));
+        if (asyncResponseTransformer == null) {
+            throw new NullPointerException("onNext must not be called with null asyncResponseTransformer");
         }
-        if (totalPartKnown.get() && part > totalParts.get()) {
-            log.info(() -> "no more parts available, stopping");
+
+        int nextPartToGet = completedParts.get() + 1;
+        if (totalPartKnown.get() && nextPartToGet > totalParts.get()) {
             subscription.cancel();
             return;
         }
-        GetObjectRequest actualRequest = this.getObjectRequest.copy(req -> req.partNumber(part));
+
+        GetObjectRequest actualRequest = getObjectRequest.copy(req -> req.partNumber(nextPartToGet));
         CompletableFuture<GetObjectResponse> future = s3.getObject(actualRequest, asyncResponseTransformer);
-        future.whenComplete((response, e) -> {
-            if (e != null) {
-                responseFuture.completeExceptionally(e);
+        future.whenComplete((response, error) -> {
+            if (error != null) {
+                onError(error);
                 return;
             }
-            completed.incrementAndGet();
-            returnResponse = response;
-            log.info(() -> String.format("received '%s'", response.contentRange()));
+            int totalComplete = completedParts.incrementAndGet();
+            log.info(() -> String.format("completed part: %s", totalComplete));
+
             Integer partCount = response.partsCount();
-            if (totalPartKnown.compareAndSet(false, true)) {
+            if (partCount != null && totalPartKnown.compareAndSet(false, true)) {
+                log.info(() -> String.format("total parts: %s", partCount));
                 totalParts.set(partCount);
                 totalPartKnown.set(true);
             }
-            log.info(() -> String.format("total parts: %s", partCount));
-            if (totalParts.get() > 1) {
+            if (partCount != null && requiresMultipart()) {
                 subscription.request(1);
```

Comment:
As we discussed offline, we need to add synchronized block on `subscription.request` and `subscription.cancel`